### PR TITLE
Trello: Inject env var for all ConfigMap values

### PR DESCRIPTION
**Pull request by mburke5678**
https://trello.com/c/RmCLchUd/460-document-downward-api-inject-env-var-for-all-configmap-values

### CODE REVIEW

**Comment by mburke5678** [comment]
@pmorie PTAL at the changes to the "Use Cases: Consuming in Environment Variable" section related to Trello Inject env var for all ConfigMap values.
https://trello.com/c/pj5Ds5bV/695-3-downward-api-inject-env-var-for-all-configmap-values

**Comment by mburke5678** [comment]
Changes are in lines 288 - 362, the _Consuming in Environment Variables_. Other changes in the file are formatting changes.   @pmorie @derekwaynecarr

File: dev_guide/configmaps.adoc [hunk 1]
```diff
@@ -307,27 +285,39 @@ objects in pods.
 [[configmaps-use-case-consuming-in-env-vars]]
 === Consuming in Environment Variables
 
-A `*ConfigMap*` can be used to populate the value of command line arguments. For
-example, consider the following `*ConfigMap*`:
+ConfigMaps can be used to populate individual environment variables or be used in
++its entirety.  As an example, consider the following ConfigMaps:
```
Comment (pmorie) [inline dev_guide/configmaps.adoc]: "or can populate environment variables from all keys that form valid environment variable names".

File: dev_guide/configmaps.adoc [hunk 2]
```diff
@@ -339,39 +329,46 @@ spec:
...
+        envFrom:<5>
+          - configMapRef:
+              name: env-config <6>
   restartPolicy: Never
 ----
+
+<1> Stanza to pull specific environment variables from a ConfigMap.
+<2> Name of the ConfigMap to pull specific environment variables.
+<3> Environment variable to pull from the ConfigMap. 
+<4> Marks the environment variable as optional and the variable will not be used.
```
Comment (pmorie) [inline dev_guide/configmaps.adoc]: "...as optional; the pod will be started even if this configmap and key do not exist"

**Review by pmorie — COMMENTED** [review]
Just a couple wording suggestions, LGTM otherwise.  Secrets can be used this way as well - is that already doc'd or does that need to be done, too?

**Comment by mburke5678** [comment]
@pmorie  There is an example in the Secrets doc  of a pod spec that uses the `secretKeyRef` parameter: "Example 4. YAML of a Pod Populating Environment Variables with Secret Data" Not as explicit as what we now have in Configmaps 
https://docs.openshift.com/container-platform/3.5/dev_guide/secrets.html#secrets-examples

**Comment by pmorie** [comment]
@mburke5678 re: secrets, okay.  I don't want to make up work for now reason, but we should apply a similar change to the secrets doc, if possible.

**Comment by mdshuai** [comment]
@mburke5678 For paul's comment he means envFrom should support secret and configmap, for secret we can do it in another pr. This LGTM, thank you.

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L1537-L1547

**Comment by mburke5678** [comment]
@openshift/team-documentation PTAL
Changes are in lines 288 - 362, the Consuming in Environment Variables.

**Comment by mburke5678** [comment]
@pmorie @mdshuai Does the `configMapRef` parameter take an `optional: true` also? If not, what happens if the specified configmap does not exist?

**Comment by mburke5678** [comment]
Bump!!
@openshift/team-documentation PTAL
Changes are in lines 288 - 362, the Consuming in Environment Variables.

File: dev_guide/configmaps.adoc [hunk 3]
```diff
@@ -307,27 +285,44 @@ objects in pods.
...
+kind: ConfigMap
+metadata:
+  name: env-config <1>
+  namespace: default
+data:
+  log_level: INFO <2>
 ----
-====
+
+<1> Name of the ConfigMap.
```
Comment (ahardin-rh) [inline dev_guide/configmaps.adoc]: `*ConfigMap*`

**Comment by ahardin-rh** [comment]
@mburke5678 Just one comment from me

**Comment by mburke5678** [comment]
[rev_history]
|xref:../dev_guide/configmaps.adoc#dev-guide-configmaps[ConfigMaps]
|Added information on injecting environment variables into pods at creation to the xref:../dev_guide/configmaps.adoc#configmaps-use-case-consuming-in-env-vars[Consuming in Environment Variable] section.
%

**Pull request merged**